### PR TITLE
Work changes

### PR DESCRIPTION
**Pull request by laerreal**
A pack of changes made during recent work.

### v4
- fixed `qdev_realize_and_unref` usage in machine boilerplate

### v3
- rework changes in `sysbusdevice.py` and `pcie.py`
- remove junk comment in `version.py`

### v2
- add new commits mainly related to device/machine boilerplate
  generation (on top of the branch)
- use `is_trace_skipped`
- fixup typos
- extend `RSPWatcher` docstring
- also adjust `size` of multiline (`QLog`) instruction

### CODE REVIEW

File: qemu/qlog.py [hunk 1]
```diff
@@ -506,6 +509,22 @@ def new_trace(self, trace, lineno):
             print("--- trace")
             print("".join(trace))
 
+        if trace[-1].startswith("Stopped execution of TB chain before 0x"):
```
Comment (BDanAnd) [inline qemu/qlog.py]: Maybe use `is_trace_skipped`?

File: debug/glv_adapter.py [hunk 2]
```diff
@@ -18,6 +18,7 @@
     range
 )
 from common import (
+    execfile,
```
Comment (BDanAnd) [inline debug/glv_adapter.py:21]: Please use `our` in this commit message.

File: debug/rsp_watcher.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            if re_symbol.match(sym):
+                yield cb, sym
+
+
+class RSPWatcher(object):
+    """ A breakpoint setting helper based on .symtab (symbol table) only.
+
+A method with such a signature is treated as a breakpoint callback:
+
+    def no_some_symbol(self)
```
Comment (BDanAnd) [inline debug/rsp_watcher.py]: `on_some_symbol` misprint. Maybe add `:` at the end.

File: qemu/makefile_patching.py [hunk 4]
```diff
@@ -18,28 +18,41 @@ def patch_makefile(mf_full_name, obj_base_name, obj_var_name, config_flag):
...
-        if lines:
-            nl_at_eof = (lines[-1][-1] == "\n")
+        # Register object file with name obj_base_name.
+        patch_line = (
+            obj_var_token
+            + " += "
+            + obj_base_name
+        )
+
+        if patch_line_n is None or patch_line == len(lines):
```
Comment (BDanAnd) [inline qemu/makefile_patching.py]: Maybe `patch_line_n == len(lines)`?

File: qemu/qlog.py [hunk 5]
```diff
@@ -395,19 +398,40 @@ def new_in_asm(self, in_asm):
...
+            # if at least one instruction of a TB overlaps another TB,
+            # the overlapping TB becomes related to new cache version.
+            if not instr.first:
+                self.tbIdMap[tb][1] = len(self.in_asm)
+
+        if instr.first:
+            self.tbIdMap[tb] = [instr.addr, len(self.in_asm)]
+
+            # It looks like, only first byte of an instruction is
+            # really needed to be accounter in tbMap.
```
Comment (BDanAnd) [inline qemu/qlog.py]: `accounter` or `accounted`?

File: qemu/qlog.py [hunk 6]
```diff
@@ -395,22 +444,41 @@ def new_in_asm(self, in_asm):
...
+                if prev_instr is None:
+                    raise RuntimeError(
+                        "in_asm record starts with an instruction tail:\n" +
+                        "\n".join(in_asm) + "\n"
+                    )
+                if DEBUG < 3:
+                    print("Join multiline instruction '%s' + '%s'" % (
+                        prev_instr, instr
+                    ))
+                prev_instr.bytes = prev_instr.bytes + instr.bytes
```
Comment (BDanAnd) [inline qemu/qlog.py]: Should  `prev_instr.size` be adjusted too?

**Review by BDanAnd — COMMENTED** [review]
v1 viewed

**Comment by laerreal** [comment]
v2 is ready

File: qemu/sysbusdevice.py [hunk 7]
```diff
@@ -553,12 +553,8 @@ def fill_source(self):
     timer_free(s->{timerN});
 """.format(timerN = self.timer_name(timerN))
 
-        self.device_unrealize = Function(
+        self.device_unrealize = Type["DeviceUnrealize"].type.use_as_prototype(
```
Comment (BDanAnd) [inline qemu/sysbusdevice.py]: What about `Type["DeviceClass"].unrealize.gen_callback(`?

File: qemu/pcie.py [hunk 8]
```diff
@@ -459,8 +457,18 @@ def fill_source(self):
...
+"""
+
+        if get_vp("use device_class_set_props"):
+            class_init_fmt += \
+                "    device_class_set_props(dc,@s{dev}_properties);\n"
+        else:
+            class_init_fmt += \
+                "    dc->props@b@b@b@b@b{pad}=@s{dev}_properties;\n"
+
+        self.class_init = Function(
```
Comment (BDanAnd) [inline qemu/pcie.py:481]:
```
used_types = [
    ...
] + (
    [ Type["device_class_set_props"] ] if get_vp(
        "use device_class_set_props"
    ) else
    []
),
```
And same in `sysbusdevice.py`.

File: qemu/version.py [hunk 9]
```diff
@@ -630,6 +630,17 @@ def define_only_qemu_2_6_0_types():
         qdev_core_header.add_type(
             Function(name = "device_class_set_parent_realize")
         )
+    # void device_class_set_props(DeviceClass *dc, Property *pr
```
Comment (BDanAnd) [inline qemu/version.py]: `void device_class_set_props(DeviceClass *dc, Property *props);` in qemu code:
1) Maybe add `)` to comment
2) Maybe rename `pr` arg to `props`
3) Maybe delete comment

File: qemu/machine.py [hunk 10]
```diff
@@ -418,31 +424,66 @@ def fill_source(self):
...
+    dev_name = dev_name
+                            )
+                        else:
+                            # TODO: test this branch using some non-sysbus
+                            #       configuration
+                            self.use_type_name("qdev_realize_and_unref")
+                            bus_name = \
+                                "BUS(%s)" % self.node_map[node.parent_bus]
+                            def_code += """\
+    qdev_realize_and_unref(@aSYS_BUS_DEVICE({dev_name}),@s{bus_name},@sNULL);
```
Comment (BDanAnd) [inline qemu/machine.py]: I think that `qdev_realize_and_unref(@a{dev_name},@s{bus_name},@sNULL);`.

File: qemu/machine.py [hunk 11]
```diff
@@ -418,31 +424,66 @@ def fill_source(self):
...
+    sysbus_realize_and_unref(@aSYS_BUS_DEVICE({dev_name}),@sNULL);
+""".format(
+    dev_name = dev_name
+                            )
+                        else:
+                            # TODO: test this branch using some non-sysbus
+                            #       configuration
+                            self.use_type_name("qdev_realize_and_unref")
+                            bus_name = \
+                                "BUS(%s)" % self.node_map[node.parent_bus]
```
Comment (BDanAnd) [inline qemu/machine.py:467]: Is `self.use_type_name("BUS")` needed?

File: qemu/machine.py [hunk 12]
```diff
@@ -418,31 +424,66 @@ def fill_source(self):
...
+    bus_name = bus_name
+                            )
+                    else:
+                        if ((node.parent_bus is None)
+                            or isinstance(node.parent_bus, SystemBusNode)
+                        ):
+                            bus_name = "NULL"
+                        else:
+                            bus_name = \
+                                "BUS(%s)" % self.node_map[node.parent_bus]
```
Comment (BDanAnd) [inline qemu/machine.py:481]: Is `self.use_type_name("BUS")` needed?

File: qemu/version.py [hunk 13]
```diff
@@ -765,12 +765,26 @@ def define_only_qemu_2_6_0_types():
             Structure("CharDriverState")
         ]
 
+    if get_vp("use QEMUChrEvent"):
```
Comment (BDanAnd) [inline qemu/version.py:767]: This enum exists since`v2.8.0-rc0` version. Maybe just add without `get_vp`?

**Review by BDanAnd — COMMENTED** [review]
v2 viewed

File: qemu/machine.py (hunk 11 above)
Comment (laerreal) [reply qemu/machine.py:467]: The variable may have type of `Bus` subclass. So, it's likely needed.

File: qemu/version.py (hunk 13 above)
Comment (laerreal) [reply qemu/version.py:767]: Existence of the enum is only matter in new qemu version.

File: qemu/version.py (hunk 9 above)
Comment (laerreal) [reply qemu/version.py]: I copied it as a hint and forget to remove...

File: qemu/machine.py (hunk 10 above)
Comment (laerreal) [reply qemu/machine.py]: dev_name is `DeviceState *` while `qdev_realize_and_unref` accepts `SysBusDevice *`. That's qemu, boy. =)

File: qemu/machine.py (hunk 11 above)
Comment (laerreal) [reply qemu/machine.py:467]: I decided to leave the text as is. Changing is not point of that PR.

**Comment by laerreal** [comment]
v3 is ready

File: qemu/machine.py (hunk 10 above)
Comment (BDanAnd) [reply qemu/machine.py]: `bool qdev_realize_and_unref(DeviceState *dev, BusState *bus, Error **errp)`
Why `SysBusDevice *`?

File: qemu/machine.py (hunk 11 above)
Comment (BDanAnd) [reply qemu/machine.py:467]: I mean that there's no indication of type `BUS` usage.

**Comment by BDanAnd** [comment]
v3 viewed

**Review by BDanAnd — COMMENTED** [review]
v3 viewed

File: qemu/machine.py (hunk 10 above)
Comment (laerreal) [reply qemu/machine.py]: Oops, I meant `sysbus_realize_and_unref`.

**Comment by laerreal** [comment]
v4 is ready

File: qemu/machine.py (hunk 11 above)
Comment (laerreal) [reply qemu/machine.py:467]: O, the question is not about that...

So. there is `self.use_type_name("BUS")` above in the code; it's needed (probably not always) and enough.

**Pull request merged**